### PR TITLE
Add a version of the parser that preserves comments

### PR DESCRIPTION
**Pull request by diml**
This should help for `dune fmt` /cc @emillon

### CODE REVIEW

**Comment by diml** [comment]
This is ready for review. The jbuild parsing is a bit messy: we only keep the location and fetch the comments afterwards, however this code should go away in 6 months, so it's not a big deal. At least it should allow to implement `dune upgrade`.

File: src/dune_lang/dune_lang.mli [hunk 1]
```diff
@@ -105,6 +105,44 @@ end with type sexp := t
 
 val add_loc : t -> loc:Loc.t -> Ast.t
 
+(** Concrete syntax tree *)
+module Cst : sig
```
Comment (emillon) [inline src/dune_lang/dune_lang.mli]: `module Concrete` would make this a bit more obvious maybe?

File: src/dune_lang/dune_lang.ml [hunk 2]
```diff
@@ -216,22 +248,47 @@ module Parser = struct
             List (loc, x :: l)
   end
 
-  let rec loop depth lexer lexbuf acc =
-    match (lexer lexbuf : Lexer.Token.t) with
+  (* To avoid writing two parsers, one for the Cst and one for the
```
Comment (emillon) [inline src/dune_lang/dune_lang.ml]: Thanks for explaining that!

File: src/dune_lang/dune_lang.mli (hunk 1 above)
Comment (diml) [reply src/dune_lang/dune_lang.mli]: While the term CST is not used as often as AST, I believe it's common enough to justify keeping it here.

File: src/dune_lang/dune_lang.ml (hunk 2 above)
Comment (diml) [reply src/dune_lang/dune_lang.ml]: No problem :)

**Pull request merged**